### ISSUE TITLE
Built-in function support

### ISSUE THREAD

**Issue by curvedmark** [curvedmark/roole#15]
- color manipulation
- image size reading
- converting image to data uri

`$darken()` or `darken()`

**Comment by curvedmark** [comment]
If use variables, then syntax should allow using mixin calling as an expression, but throw in evaluator when the variable is a mixin.

If use keywords, likely to collide with CSS functions, and hard to spot which ones are Roole specific.

If go with variables, maybe should also allow users to define mixins that can return a value:

```
$mixin = @mixin $value
    @return $value + 1
```

This is no longer a mixin, the name is misleading, but is it really matter? Or simply change `@mixin` to `@function` and be vague on the concept? Or provide both `@mixin` and `@function`?

Another solution is to disallow the current mixin calling syntax, use the syntax suggested in #17, and change `@mixin` to `@function`:

```
$func = @function $value
    property: $value
    @return $value

foo
    @call $func bar
    property2: $func(qux)
```

compiles to 

```
foo {
    property: bar;
    property2: qux;
}
```

replace `@call` with `@mixin` might make it clearer.

**Comment by jamiebuilds** [comment]
I'd prefer if this worked like SASS:

``` sass
selector
    property: function(options)
```

It's more concise and easier to follow

**Comment by curvedmark** [comment]
I don't feel like overloading CSS functions. Just by looking at the code, it's not clear it's defined by the user or it is a CSS function. Using a variable makes it clear.

**Comment by jamiebuilds** [comment]
You could always tack a character before it. ie. `+function()`, `@function()`, or `$function()`

**Comment by iamcarrico** [comment]
I am with @thejameskyle that this should be concise... and possibly just user a character in front. 

So, in Sass the difference between the two is that a mixin will display the CSS, while a function will only return a variable -- that can be either displayed like in the @thejameskyle's example above-- or stored as a variable. Although I like that separation... It isn't entirely necessary. If we just allow an @return happen within the current @mixin support... and ensure the return can be stored as a new variable... I think this would be set.

**Comment by curvedmark** [comment]
> I am with @thejameskyle that this should be concise... and possibly just user a character in front.

The syntax `$function()` suggested by @thejameskyle is the same as the planned syntax (calling a function stored in a variable will return it's return value)

> If we just allow an `@return` happen within the current `@mixin` support... and ensure the return can be stored as a new variable... I think this would be set.

Agreed, and that's what i was talking about in the previous post.

To make it clear, I will sum up the options here, using the new CSS syntax:

### Syntax 1

Define mixins and functions differently, but call them in the same way.

Mixin:

```
$foo = @mixin $param1, $param2 {
    margin: 0;
};

body {
    $foo();
}
```

Function:

```
$bar = @function $param1, $param2 {
    @return 1px * 2;
};

body {
    margin: $bar();
}
```

Pros:
- Clear separation of concepts. `@mixin` returns rules inside it, `@function` returns values.
- Easy to implement, parser can quickly throw syntax error if `@return`s are specified in `@mixin`, or rules (e.g., rulesets, medias) are specified in `@function`, so evaluator don't have to worry about how to ignore them.

Cons:
- Calling them the same way makes the calling code ambiguous. For example, in the first snippet, without the definition of  `$foo`, it's not clear whether you are calling a mixin or a function.

### Syntax 2

Basically the same as syntax 1, but using syntax suggested by @thejameskyle in #44.

Mixin:

```
@mixin $foo($param1, $param2) {
    margin: 0;
}

body {
    $foo(a, b);
}
```

Function:

```
@function $bar($param1, $param2) {
    @return 1px * 2;
}

body {
    margin: $bar(a, b);
}
```

Pros:
- Same pros in syntax 1
- People from a Sass background will feel right at home

Cons:
- Same cons in syntax 1
- Does not make it extremely clear that mixins and functions are just regular values which can be passed directly as arguments.

### Syntax 3

Define mixins and functions in the same way, but call them differently.

Mixin:

```
$foo = @function $param1, $param2 {
    margin: 0;
};

body {
    @mixin $foo a, b;
}
```

Function:

```
$bar = @function $param1, $param2 {
    @return 1px * 2;
};

body {
    margin: $bar(a, b);
}
```

Pros:
- Readability is good, when looking at the definition or the calling code. It's very clear whether it works like a mixin or a function (especially `@mixin $foo`, which clearly says it is mixing in rules stored in `$foo`) .

Cons:
- Not sure about the arguments syntax when calling a mixin. For example, with `@mixin $foo $bar, $baz`, it's very likely people will write `@mixin $foo, $bar, $baz` (notice the first comma), the last two variables are arguments, the first one isn't, so the first comma really shouldn't be there. Allowing `@mixin $foo(a, b)` is confusing, it looks like mixing in a mixin returned by the function call.
- Not easy to implement, the evaluator need to ignore rules (e.g., rulesets, medias) when the variable is called as a function, and ignore `@return`s when the variable is called as a mixin.

### Syntax 4

Define mixins and functions in the same way, and call them in the same way.

I won't bother with this one. Too ambiguous.

**Comment by iamcarrico** [comment]
Honestly--- I could care less between 1, 2, and 3--- and I would opt for what would be easiest to maintain. I dont think any of them will cause any heartache between them.

The one thing I realized last night that is important to try and do--- is allow the calling of certain functions that are defined in JavaScript outside of the .roo files (possibly within an extension as in #67 ), would be awesome. This might hurt the ability of having functions be named similar to a variable--- and might cause some issues.

**Comment by curvedmark** [comment]
Landed in 76eb2ea6f9677a4ea51df6db7ee5b340c31fa194
